src/ring_jwt_middleware/core.clj

```diff
@@ -233,68 +276,85 @@
+  - pubkey-fn ; should contain a function that once called will return the public key
+  - is-revoked-fn ; should be a function that takes a decoded jwt and return true if the jwt is revoked
+  - jwt-check-fn ; should be a function taking a raw JWT string, and a decoded JWT and returns a list of errors or nil if no error is found.
+  - jwt-max-lifetime-in-sec ; maximal lifetime of a JWT in seconds (takes priority over :exp)
+  - post-jwt-format-fn ; a function taking a JWT and returning a data structure representing the identity of a user
+
+  "
   [{:keys [pubkey-path
            pubkey-fn
            is-revoked-fn
            jwt-check-fn
            jwt-max-lifetime-in-sec
-           post-jwt-format-fn
-           no-jwt-handler
-           error-handler
-           structured-log-fn]
+           post-jwt-format-fn]
     :or {jwt-max-lifetime-in-sec default-jwt-lifetime-in-sec
          is-revoked-fn no-revocation-strategy
-         post-jwt-format-fn jwt->user-id
-         no-jwt-handler forbid-no-jwt-header-strategy
-         structured-log-fn default-structured-log}}]
+         post-jwt-format-fn jwt->user-id}} :- Config]
   (let [p-fn (or pubkey-fn (constantly (public-key pubkey-path)))
         is-revoked-fn (if (fn? is-revoked-fn)
                         is-revoked-fn
-                        (do (structured-log-fn
-                             "is-revoked-fn is not a function! no-revocation-strategy is used."
-                             {:level :error})
-                            no-revocation-strategy))
-        handle-error (or error-handler default-error-handler)]
+                        (do (log/error "is-revoked-fn is not a function! no-revocation-strategy is used.")
+                            no-revocation-strategy))]
+    (fn [handler]
+      (fn [request]
+        (let [authentication-result
+              (let-either [raw-jwt (get-jwt request)
+                           {:keys [jwt] :as  _decoded-result} (decode raw-jwt p-fn)
+                           _ (validate-jwt raw-jwt jwt jwt-max-lifetime-in-sec jwt-check-fn)
+                           _ (try (if (is-revoked-fn jwt)
+                                    (->err :jwt_revoked "JWT is revoked" {:jwt jwt})
+                                    (->pure :ok))
+                                  (catch Exception e
+                                    (->err :jwt-revocation-fn-exception
+                                           "is-revoked-fn thrown an exception"
+                                           {:level :error
+                                            :exception e
+                                            :jwt jwt})))]
+                (->pure {:identity (post-jwt-format-fn jwt)
+                         :jwt jwt}))]
+          (handler (into request (<-result authentication-result))))))))
+
+(defn mk-wrap-authorization
+  "A function building a middleware taking care of the authorization logic.
```

Comment:
Can you help me understand what this function has to do with authorization? All I see are error handlers and special case handling.

Reply:
The authentication will just take care of adding a few fields to the ring request.

The authorization is responsible to make a call fail or not if the request contains an authentication error or not.

Generally the authorization logic should not be the concern of this lib but external.
Still, there is a best practice we could enforce by making the distinction between request without any authentication, and request with failed authentication attempts.

For the second case, the request should fail, and this is enforced by using a function without any `handler` as an argument.